src/printer-markdown.js

```diff
@@ -0,0 +1,434 @@
+
+const docBuilders = require("./doc-builders");
+const concat = docBuilders.concat;
+const join = docBuilders.join;
+const line = docBuilders.line;
+const hardline = docBuilders.hardline;
+const fill = docBuilders.fill;
+const align = docBuilders.align;
+const docPrinter = require("./doc-printer");
+const printDocToString = docPrinter.printDocToString;
+const escapeStringRegexp = require("escape-string-regexp");
+
+const SINGLE_LINE_NODE_TYPES = ["heading", "tableCell", "footnoteDefinition"];
+
+const SIBLING_NODE_TYPES = ["listItem", "definition", "footnoteDefinition"];
+
+const INLINE_NODE_TYPES = [
+  "inlineCode",
+  "emphasis",
+  "strong",
+  "delete",
+  "link",
+  "linkReference",
+  "image",
+  "imageReference",
+  "footnote",
+  "footnoteReference",
+  "sentence",
+  "whitespace",
+  "word"
+];
+
+const INLINE_NODE_WRAPPER_TYPES = INLINE_NODE_TYPES.concat([
+  "tableCell",
+  "paragraph"
+]);
+
+function genericPrint(path, options, print) {
+  const node = path.getValue();
+
+  switch (node.type) {
+    case "root":
+      return normalizeDoc(
+        concat([printChildren(path, options, print), hardline])
+      );
+    case "paragraph":
+      return printChildren(path, options, print, {
+        postprocessor: fill
+      });
+    case "sentence":
+      return printChildren(path, options, print);
+    case "word":
+      return escapeString(node.value, ["_", "*", "~~"]);
+    case "whitespace": {
+      return hasParentType(path, SINGLE_LINE_NODE_TYPES) ? " " : line;
+    }
+    case "emphasis":
+      return concat(["*", printChildren(path, options, print), "*"]);
+    case "strong":
+      return concat(["**", printChildren(path, options, print), "**"]);
+    case "delete":
+      return concat(["~~", printChildren(path, options, print), "~~"]);
+    case "inlineCode":
+      return concat(["`", node.value, "`"]);
+    case "link":
+      return concat([
+        "[",
+        printChildren(path, options, print),
+        "](",
+        node.url,
+        node.title ? ` "${node.title}"` : "",
+        ")"
+      ]);
+    case "image":
+      return concat([
+        "![",
+        node.alt || "",
+        "](",
+        node.url,
+        node.title ? ` "${node.title}"` : "",
+        ")"
+      ]);
+    case "blockquote":
+      return concat(["> ", align("> ", printChildren(path, options, print))]);
+    case "heading":
+      return concat([
+        "#".repeat(node.depth) + " ",
+        printChildren(path, options, print)
+      ]);
+    case "code":
+      return concat([
+        "```",
+        node.lang || "",
+        hardline,
+        node.value,
+        hardline,
+        "```"
+      ]);
+    case "yaml":
+      return concat(["---", hardline, node.value, hardline, "---"]);
+    case "html":
+      return node.value;
+    case "list":
+      return printChildren(path, options, print, {
+        processor: (childPath, index) => {
+          const prefix = node.ordered ? `${node.start + index}. ` : "- ";
```

Comment:
why include the index? One of the things I like about markdown is that I can move around the entries in an ordered list without having to renumber them

(loving this PR btw, prettier keeps on making new stuff pretty :D)

Reply:
Well, I think it looks better with the index and the index will be renumbered by Prettier, you don't have to edit it manually.

Reply:
so

```md
1. one
3. two
2. three
```

is changed to

```md
1. one
2. two
3. three
```
?

In that case I guess it's fine 🙂

Reply:
Yes, only the initial number matters.

Reply:
![](http://g.recordit.co/MAtanZMZRo.gif) awesome!